File: 4.8-bc/ts/stage-3-hexagonal/src/billing/app/service.ts
// Billing use case — оформлення підписки.
import { randomUUID } from "node:crypto";

import type { EventBus } from "../../shared/events/bus.js";
import { InvalidPlanError } from "../domain/errors.js";
import type { Repository } from "../domain/repository.js";
import {
  SubscriptionCreated,
  type Subscription,
} from "../domain/subscription.js";

const VALID_PLANS = new Set(["basic", "pro", "enterprise"]);

export class Service {
  constructor(
    private readonly repo: Repository,
    private readonly bus: EventBus,
  ) {}

  async subscribe(userId: string, plan: string): Promise<Subscription> {
    if (!VALID_PLANS.has(plan)) throw new InvalidPlanError();

    const now = new Date();
    const next = new Date(now);
    next.setMonth(next.getMonth() + 1);
    const sub: Subscription = {
      id: randomUUID(),
      userId,
      plan,
      nextChargeAt: next,
      createdAt: now,
    };
    await this.repo.create(sub);
    await this.bus.publish(
      new SubscriptionCreated(sub.id, sub.userId, sub.plan, sub.createdAt),
    );
    return sub;
  }
}

File: 4.8-bc/py/stage-3-hexagonal/app/billing/app/service.py
"""Billing use cases — Subscribe.

The plan whitelist lives here, in the use case layer. Adding a plan is a
business decision that an admin / product manager owns.
"""
from __future__ import annotations

from datetime import UTC, datetime, timedelta
from uuid import UUID, uuid4

from app.billing.domain.repository import Repository
from app.billing.domain.subscription import (
    InvalidPlanError,
    Subscription,
    SubscriptionCreated,
)
from app.shared.events import EventBus

VALID_PLANS = frozenset({"basic", "pro", "enterprise"})


class Service:
    def __init__(self, repo: Repository, bus: EventBus) -> None:
        self._repo = repo
        self._bus = bus

    async def subscribe(self, user_id: UUID, plan: str) -> Subscription:
        if plan not in VALID_PLANS:
            raise InvalidPlanError()
        now = datetime.now(UTC)
        sub = Subscription(
            id=uuid4(),
            user_id=user_id,
            plan=plan,
            next_charge_at=now + timedelta(days=30),
            created_at=now,
        )
        await self._repo.create(sub)
        await self._bus.publish(
            SubscriptionCreated(
                subscription_id=sub.id,
                user_id=sub.user_id,
                plan=sub.plan,
                at=sub.created_at,
            )
        )
        return sub
